### DISCUSSION TITLE
system include extraction failing

### DISCUSSION THREAD

**Discussion by mellery451** [clangd/clangd#1473] [Q&A] [answered]
I have a situation with a toolchain where the system include extraction fails..in the log I see:

       System include extraction: start marker not found: <details redacted>

..and indeed when I run the command that it seems to be running, I get basically empty output:

    # 1 "<stdin>"
    # 1 "<built-in>"
    # 1 "<command-line>"
    # 1 "<stdin>"

What else can I do to try to diagnose this issue? Not sure what would cause the toolchain to fail to give anythiing useful about includes (it's a gcc based toolchain for qnx in this case).

**Comment by HighCommander4** [comment]
> when I run the command that it seems to be running

Can you quote the command as well?

**Comment by mellery451** [comment]
here's the log message:

```
System include extraction: start marker not found: cc: looking for gcc_ntox86_64 in external/qnx_tools_qnx710_ga1/host/linux/x86_64/etc/qcc/gcc/8.3.0/gcc_ntox86_64++.conf\ncc: looking for gcc_ntox86_64 in external/qnx_tools_qnx710_ga1/host/linux/x86_64/etc/qcc/gcc/8.3.0/gcc_ntox86_64.conf\nexternal/qnx_tools_qnx710_ga1/host/linux/x86_64/usr/lib/gcc/x86_64-pc-nto-qnx7.1.0/8.3.0/cc1plus -E -quiet -nostdinc -nostdinc++ -std=gnu++11 -D__QNX__ -D__QNXNTO__ -D__GNUC__=8 -D__GNUC_MINOR__=3 -D__GNUC_PATCHLEVEL__=0 -D__NO_INLINE__ -D__DEPRECATED -D__unix__ -D__unix -D__ELF__ -fPIE -D__X86_64__ -D__LITTLEENDIAN__ -Asystem(unix) -isystem external/qnx_tools_qnx710_ga1/target/qnx7/usr/include/c++/v1 -isystem external/qnx_tools_qnx710_ga1/target/qnx7/usr/include -isystem external/qnx_tools_qnx710_ga1/host/linux/x86_64/usr/lib/gcc/x86_64-pc-nto-qnx7.1.0/8.3.0/include - -o -\ncc: removing  /tmp/fileWQGIAe_qcc_3826503\n\n"
```

..and TBH, I'm confused about how it came up with some of these flags (the c++11 standard for example). I'm using a compile database and the flags in there should generally be c++14 ish

**Reply by HighCommander4** [reply]
Yeah, that doesn't look right. There should be flags there that aren't (like `-v`), and there are flags there that shouldn't be (like the `-D` flags, I don't think we propagate those into the extraction command).

Woulf you be able to share complete clangd logs taken with `--log=verbose`?

**Reply by mellery451** [reply]
unfortunately I'm working with closed source and most of the stuff in the logs I probably can't share. FWIW the startup message from clangd is at https://gist.githubusercontent.com/mellery451/087d80ec90566f13de591f550963ea11/raw/71ae68e181e6c441629921e6ffb7783ddbd77976/foo.json . If you can tell me what else I might be looking for, I can try to extract bits of the log info. thanks.

**Reply by HighCommander4** [reply]
Understood.

I also realized I misinterpreted the previous log message: what's printed after `start marker not found:` is the **output** of the extraction command. So it looks like the driver in your case in some sort of wrapper script, and it's printing something very different from what we expect from a gcc-compatible driver.

Also I thought that clangd would print the extraction command itself somewhere in the logs, but it doesn't look like it does, so let's infer it instead.

Can you look in the logs for `ASTWorker building file <filename> version <version> with command` when you open `<filename>` in your editor, and share the command that's printed on the subsequent line (or as much of it as you can)? Particularly important in this case are the executable at the beginning of the command, and any `--target`, `--sysroot`, or `-nostdinc` flags present in the command.

**Reply by mellery451** [reply]
is there  a way to see the actual command that run to extract info? MY compile database has commands that all start with:'

    external/qnx_tools_qnx710_ga1/bin/q++

which is just a symlink to qcc in the same dir, which is very much an elf binary (not a script of some sort)...so I'm not sure how a wrapper if getting in the way.  I DO have a wrapper script around my clangd startup (called "myclangd") that I use to set a few env vars required for QNX...which I was hoping would help with the extraction command, but maybe that's not working. I can't tell.

**Reply by mellery451** [reply]
as for the AST message, I think this is what we have:

[ERROR][2023-01-23 11:38:40] .../vim/lsp/rpc.lua:733	"rpc"	"/home/mellery/bin/myclangd"	"stderr"	"V[11:38:40.142]
Ignored diagnostic. argument unused during compilation: '-MF =<long path redacted>\n
V[11:38:40.145] Driver produced command:
cc1
-cc1
-triple x86_64-unknown-linux-gnu
-fsyntax-only
-disable-free
-clear-ast-before-backend
-disable-llvm-verifier
-discard-value-names
-main-file-name <redacted>
-mrelocation-model pic
-pic-level 2
-pic-is-pie
-mframe-pointer=all
-fmath-errno
-ffp-contract=on
-fno-rounding-math
-mconstructor-aliases
-funwind-tables=2
-target-cpu x86-64
-tune-cpu generic
-mllvm
-treat-scalable-fixed-error-as-warning
-debugger-tuning=gdb
-ffunction-sections
-fdata-sections
-fcoverage-compilation-dir=/home/mellery/work/ndas
-resource-dir /home/linuxbrew/.linuxbrew/Cellar/llvm/15.0.7_1/lib/clang/15.0.7
-D _POSIX_C_SOURCE=200112L
-D _QNX_SOURCE
-D FD_SETSIZE=1024
-D NDEBUG
-D _FORTIFY_SOURCE=1
-D <misc custom macros>
<hundreds of include paths>

I don't see any sysroot arg...hmm

also, compile command is gigantic (we have hundreds of long include paths...thanks you bazel)...and appears to be truncated in the log. I hope that's just a logging thing and doesn't mean the command itself was truncated...could be that sysroot appeared somewhere after the truncation.

**Reply by HighCommander4** [reply]
> as for the AST message, I think this is what we have:

I'm actually looking for the line _before_ the "Driver produced command" line, which prints a command similar to the one from `compile_commands.json`, with some minor modifications. Based on your other comment, I would expect it to start with `external/qnx_tools_qnx710_ga1/bin/q++`, I just wanted to verify that's the case.

> I don't see any sysroot arg...hmm

That's fine, it's not required (and the common case is not to have it).

**Reply by HighCommander4** [reply]
> Based on your other comment, I would expect it to start with `external/qnx_tools_qnx710_ga1/bin/q++`, I just wanted to verify that's the case.

Anyways, assuming that's the case, the extraction command would be `external/qnx_tools_qnx710_ga1/bin/q++ -E -x c++ - -v` -- can you try running that on the command line and see what it outputs?

**Reply by mellery451** [reply]
```
cc: looking for gcc_ntox86_64 in external/qnx_tools_qnx710_ga1/host/linux/x86_64/etc/qcc/gcc/8.3.0/gcc_ntox86_64++.conf
cc: looking for gcc_ntox86_64 in external/qnx_tools_qnx710_ga1/host/linux/x86_64/etc/qcc/gcc/8.3.0/gcc_ntox86_64.conf
external/qnx_tools_qnx710_ga1/host/linux/x86_64/usr/lib/gcc/x86_64-pc-nto-qnx7.1.0/8.3.0/cc1plus -E -quiet -nostdinc -nostdinc++ -std=gnu++11 -D__QNX__ -D__QNXNTO__ -D__GNUC__=8 -D__GNUC_MINOR__=3 -D__GNUC_PATCHLEVEL__=0 -D__NO_INLINE__ -D__DEPRECATED -D__unix__ -D__unix -D__ELF__ -fPIE -D__X86_64__ -D__LITTLEENDIAN__ -Asystem(unix) -isystem external/qnx_tools_qnx710_ga1/target/qnx7/usr/include/c++/v1 -isystem external/qnx_tools_qnx710_ga1/target/qnx7/usr/include -isystem external/qnx_tools_qnx710_ga1/host/linux/x86_64/usr/lib/gcc/x86_64-pc-nto-qnx7.1.0/8.3.0/include - -o -
# 1 "<stdin>"
# 1 "<built-in>"
# 1 "<command-line>"
# 1 "<stdin>"
cc: removing  /tmp/filevFc2AN_qcc_3982537
```

**Reply by mellery451** [reply]
it's just stunningly empty...and I tried the same basic command with my other toolchain (linaro gcc) and it produces plenty of output. this qnx toolchain is just being difficult.

**Reply by HighCommander4** [reply]
Ok, yeah, that matches what we saw before, and does not contain standard library includes.

By comparison, here is what `g++ -E -x c++ - -v` otuputs for me:

```
$ g++ -E -x c++ - -v
Using built-in specs.
COLLECT_GCC=g++
OFFLOAD_TARGET_NAMES=nvptx-none
OFFLOAD_TARGET_DEFAULT=1
Target: x86_64-linux-gnu
Configured with: ../src/configure -v --with-pkgversion='Debian 8.3.0-6' --with-bugurl=file:///usr/share/doc/gcc-8/README.Bugs --enable-languages=c,ada,c++,go,brig,d,fortran,objc,obj-c++ --prefix=/usr --with-gcc-major-version-only --program-suffix=-8 --program-prefix=x86_64-linux-gnu- --enable-shared --enable-linker-build-id --libexecdir=/usr/lib --without-included-gettext --enable-threads=posix --libdir=/usr/lib --enable-nls --enable-bootstrap --enable-clocale=gnu --enable-libstdcxx-debug --enable-libstdcxx-time=yes --with-default-libstdcxx-abi=new --enable-gnu-unique-object --disable-vtable-verify --enable-libmpx --enable-plugin --enable-default-pie --with-system-zlib --with-target-system-zlib --enable-objc-gc=auto --enable-multiarch --disable-werror --with-arch-32=i686 --with-abi=m64 --with-multilib-list=m32,m64,mx32 --enable-multilib --with-tune=generic --enable-offload-targets=nvptx-none --without-cuda-driver --enable-checking=release --build=x86_64-linux-gnu --host=x86_64-linux-gnu --target=x86_64-linux-gnu
Thread model: posix
gcc version 8.3.0 (Debian 8.3.0-6)
COLLECT_GCC_OPTIONS='-E' '-v' '-shared-libgcc' '-mtune=generic' '-march=x86-64'
 /usr/lib/gcc/x86_64-linux-gnu/8/cc1plus -E -quiet -v -imultiarch x86_64-linux-gnu -D_GNU_SOURCE - -mtune=generic -march=x86-64
ignoring duplicate directory "/usr/include/x86_64-linux-gnu/c++/8"
ignoring nonexistent directory "/usr/local/include/x86_64-linux-gnu"
ignoring nonexistent directory "/usr/lib/gcc/x86_64-linux-gnu/8/../../../../x86_64-linux-gnu/include"
#include "..." search starts here:
#include <...> search starts here:
 /usr/include/c++/8
 /usr/include/x86_64-linux-gnu/c++/8
 /usr/include/c++/8/backward
 /usr/lib/gcc/x86_64-linux-gnu/8/include
 /usr/local/include
 /usr/lib/gcc/x86_64-linux-gnu/8/include-fixed
 /usr/include/x86_64-linux-gnu
 /usr/include
End of search list.
```

So, unfortunately, this is a case of `q++`'s driver syntax not being sufficiently compatible with `g++` to use with clangd's `--query-driver` feature.

Some possible ways forward:

 * Use a different mechanism than `--query-driver` to provide clangd with `q++`'s built-in include paths (for example, specify them manually in a [clangd config file](https://clangd.llvm.org/config.html) with `CompileFlags:` `Add:`.
 * Hack together a script that prints the built-in include paths in a gcc-like format and use that as your compile command's argv[0]. (In clangd 16, you can override the argv[0] with `CompileFlags:` `Compiler:` in the config file, without having to change the `compile_commands.json`.)
 * If we can figure out how to extract the built-in includes from `q++`, consider teaching clangd how to handle `q++`

**Reply by mellery451** [reply]
looking at the docs for this compiler, looks like `-vv` produces more output than just `-v`:

```
-cc: looking for gcc_ntoaarch64le in external/qnx_tools_qnx710_ga1/host/linux/x86_64/etc/qcc/gcc/8.3.0/gcc_ntoaarch64le++.conf
cc: looking for gcc_ntoaarch64le in external/qnx_tools_qnx710_ga1/host/linux/x86_64/etc/qcc/gcc/8.3.0/gcc_ntoaarch64le.conf
external/qnx_tools_qnx710_ga1/host/linux/x86_64/usr/lib/gcc/aarch64-unknown-nto-qnx7.1.0/8.3.0/cc1plus -v -E -quiet -nostdinc -nostdinc++ -std=gnu++11 -D__QNX__ -D__QNXNTO__ -D__GNUC__=8 -D__GNUC_MINOR__=3 -D__GNUC_PATCHLEVEL__=0 -D__NO_INLINE__ -D__DEPRECATED -D__unix__ -D__unix -D__ELF__ -fPIE -D__LITTLEENDIAN__ -Asystem(unix) -isystem external/qnx_tools_qnx710_ga1/target/qnx7/usr/include/c++/v1 -isystem external/qnx_tools_qnx710_ga1/target/qnx7/usr/include -isystem external/qnx_tools_qnx710_ga1/host/linux/x86_64/usr/lib/gcc/aarch64-unknown-nto-qnx7.1.0/8.3.0/include - -o -
#include "..." search starts here:
#include <...> search starts here:
 external/qnx_tools_qnx710_ga1/target/qnx7/usr/include/c++/v1
 external/qnx_tools_qnx710_ga1/target/qnx7/usr/include
 external/qnx_tools_qnx710_ga1/host/linux/x86_64/usr/lib/gcc/aarch64-unknown-nto-qnx7.1.0/8.3.0/include
End of search list.
# 1 "<stdin>"
# 1 "<built-in>"
# 1 "<command-line>"
# 1 "<stdin>"
```
..still not a ton of output, but at least there are a few include paths. Is there any way to customize these extraction flags?

**Reply by mellery451** [reply, marked as answer]
I'm a fool - there is a g++ compatible wrapper already in this toolchain dir and it handles the standard flags properly. I'll just change my database or use the Compiler config workaround that you suggest. thanks for all your help
